diff -up ntp-4.2.4p8/configure.editline ntp-4.2.4p8/configure
--- ntp-4.2.4p8/configure.editline	2009-12-08 14:07:43.000000000 +0100
+++ ntp-4.2.4p8/configure	2009-12-08 17:51:52.000000000 +0100
@@ -17216,9 +17216,15 @@ fi
 done
 
 
+READLINE_LIBS="-ledit"
+cat >>confdefs.h <<\_ACEOF
+#define HAVE_LIBEDIT
+_ACEOF
+
 # following block becomes on 4.2.5: NTP_LINEEDITLIBS
 for ac_header in readline/history.h readline/readline.h
 do :
+  break
   as_ac_Header=`$as_echo "ac_cv_header_$ac_header" | $as_tr_sh`
 ac_fn_c_check_header_mongrel "$LINENO" "$ac_header" "$as_ac_Header" "$ac_includes_default"
 eval as_val=\$$as_ac_Header
@@ -17232,7 +17238,7 @@ fi
 done
 
 case "$ac_cv_header_readline_history_h$ac_cv_header_readline_readline_h" in
- *no*) ;;
+ *) ;;
  *) save_LIBS=$LIBS
     LIBS=
     # Ralf Wildenhues: either unset ... or cache READLINE_LIBS
diff -up ntp-4.2.4p8/ntpdc/ntpdc.c.editline ntp-4.2.4p8/ntpdc/ntpdc.c
--- ntp-4.2.4p8/ntpdc/ntpdc.c.editline	2009-12-08 12:55:59.000000000 +0100
+++ ntp-4.2.4p8/ntpdc/ntpdc.c	2009-12-08 17:50:14.000000000 +0100
@@ -26,9 +26,8 @@
 # define closesocket close
 #endif /* SYS_WINNT */
 
-#if defined(HAVE_LIBREADLINE) || defined (HAVE_LIBEDIT)
-# include <readline/readline.h>
-# include <readline/history.h>
+#if defined (HAVE_LIBEDIT)
+# include <editline/readline.h>
 #endif /* HAVE_LIBREADLINE || HAVE_LIBEDIT */
 
 #ifdef SYS_VXWORKS
diff -up ntp-4.2.4p8/ntpq/ntpq.c.editline ntp-4.2.4p8/ntpq/ntpq.c
--- ntp-4.2.4p8/ntpq/ntpq.c.editline	2009-12-08 12:55:59.000000000 +0100
+++ ntp-4.2.4p8/ntpq/ntpq.c	2009-12-08 17:50:14.000000000 +0100
@@ -30,9 +30,8 @@
 # define closesocket close
 #endif /* SYS_WINNT */
 
-#if defined(HAVE_LIBREADLINE) || defined(HAVE_LIBEDIT)
-# include <readline/readline.h>
-# include <readline/history.h>
+#if defined(HAVE_LIBEDIT)
+# include <editline/readline.h>
 #endif /* HAVE_LIBREADLINE || HAVE_LIBEDIT */
 
 #ifdef SYS_VXWORKS
